src/autorest-core/lib/approved-imports/multi-promise.ts

```diff
@@ -0,0 +1,173 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+/**
+ * Why not observables or on "data"&"end" pattern?
+ * 
+ * - syntax support of MultiPromise
+ *   - you can `await` to consume them
+ *   - you can use `async` methods to create them
+ * - "data"&"end" pattern has drawbacks
+ *   - on consumer side
+ *     - single occurrence of "end" purely contractual and NOT guaranteed by type system! => pot. source of bugs
+ *     - if all data is required for processing, it has to be collected manually (introduces mutable state)
+ *   - on producer side
+ *     - event subscription system requires mutable state (list of subscribers)
+ *   - in general
+ *     - can a consumer unsubscribe? how? (sounds like further state!)
+ *     - can a consumer miss some events by subscribing "too late", i.e. after an event was fired?
+ *       - if yes: induces management complexity (all pipeline steps have to be hooked up FULLY before starting any computation)
+ *       - if no: so you have to buffer everything. even after "end" was fired, someone may subscribe...
+ * - in contrast, MultiPromise provides all that for free
+ *   - immutable data structure, no explicit state
+ *   - type system enforces order of events and single end of things
+ *   - minimum amount of "past events" stored due to single linked list strucutre:
+ *     - whoever holds a MultiPromise prevents garbage collection
+ *     - consumption of an event effectively decrements ref-count on associated node
+ *   - leverages error infrastrucure of Promise
+ * - easy to "marshal" from/to any other paradigm (events, streams, regular promises, ...)
+ */
+
+
+export type MultiPromiseItem<T> = { readonly current: T, readonly next: MultiPromise<T> };
```

Comment:
thinking of "customizable pipeline" I thought that modules (like the modeler, the generator, the validator, ...) have those guys as (named) input and output channels, connected merely by the config file (while currently this is hard coded into some function). But not sure if that's the way to go